src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/NhsAckProducerService.java

```diff
@@ -0,0 +1,157 @@
+package uk.nhs.digital.nhsconnect.lab.results.inbound;
+
+import com.github.mustachejava.Mustache;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import uk.nhs.digital.nhsconnect.lab.results.mesh.message.WorkflowId;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.Interchange;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.NhsAckContent;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.NhsAckStatus;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.NhsAckMessageContent;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.InterchangeHeader;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.MessageHeader;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.message.InterchangeParsingException;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.message.MessageParsingException;
+import uk.nhs.digital.nhsconnect.lab.results.sequence.SequenceService;
+import uk.nhs.digital.nhsconnect.lab.results.utils.TimestampService;
+import uk.nhs.digital.nhsconnect.lab.results.inbound.MessageProcessingResult.Success;
+import uk.nhs.digital.nhsconnect.lab.results.inbound.MessageProcessingResult.Error;
+
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.List;
+
+import static uk.nhs.digital.nhsconnect.lab.results.utils.TemplateUtils.loadTemplate;
+import static uk.nhs.digital.nhsconnect.lab.results.utils.TemplateUtils.fillTemplate;
+
+@Component
+@RequiredArgsConstructor(onConstructor_ = @Autowired)
+public class NhsAckProducerService {
+
+    private static final String MESSAGE_ACCEPTED_STATUS_CODE = "MA";
+    private static final String MESSAGE_REJECTED_STATUS_CODE = "MR";
+
+    private static final int MESSAGE_COMMON_SEGMENT_COUNT = 5;
+
+    private final TimestampService timestampService;
+
+    private final SequenceService sequenceService;
+
+    private Mustache nhsAckTemplate = loadTemplate("nhsAck.mustache");
+
+    public String createNhsAckEdifact(WorkflowId workflowId, Interchange interchange,
+                                      List<MessageProcessingResult> messageProcessingResults) {
+
+        InterchangeHeader interchangeHeader = interchange.getInterchangeHeader();
+
+        var nhsAckContent = NhsAckContent.builder()
+            .interchangeSender(interchangeHeader.getSender())
```

Comment:
I think this should be swapped. When we receive interchange it has sender (them) and recipient (us). When sending back an NHSACK, we should be using sender (us) and recipient (them)
Check other places

Reply:
I've made this change.